### PR TITLE
Support for Time and Date Datatypes

### PR DESCRIPTION
**Pull request by kaimal11**
includes date, time, dateTime, and dateTimeStamp datatypes

### CODE REVIEW

File: src/rdf4cpp/rdf/datatypes/xsd/Date.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        //default values
+        tm.tm_sec = 0;     // seconds of minutes from 0 to 61
+        tm.tm_min = 0;     // minutes of hour from 0 to 59
+        tm.tm_hour = 0;    // hours of day from 0 to 24
+        tm.tm_isdst = -1;  // daylight saving
+
+        return mktime(&tm);
+    } else {
+        throw std::runtime_error("XSD Parsing Error");
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/Date.hpp:52]: The standard defines the "day-of-month values" constraint (e.g., April has 30 days, so 31 is not a valid value). We have implement this constraint here.

File: src/rdf4cpp/rdf/datatypes/xsd/Date.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * Specialisation of from_string template function.
+ */
+template<>
+inline LiteralDatatypeImpl<xsd_date>::cpp_type LiteralDatatypeImpl<xsd_date>::from_string(std::string_view s) {
+
+    const std::regex date_regex(R"(-?([1-9][0-9]{3,}|0[0-9]{3})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?)");
+
+    if (std::regex_match(s.data(), date_regex)) {
+        struct tm tm {};
+        strptime(s.data(), "%Y-%m-%d", &tm);
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/Date.hpp:42]: How do you deal with the timezone fragment (which is optionally provided)?

File: src/rdf4cpp/rdf/datatypes/xsd/Date.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+/**
+ * Specialisation of to_string template function.
+ */
+template<>
+inline std::string LiteralDatatypeImpl<xsd_date>::to_string(const cpp_type &value) {
+
+    char str[32];
+    std::strftime(str, 32, "%Y-%m-%d", std::localtime(&value));
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/Date.hpp:65]: We also have to consider how to treat the timezone fragment here.

File: tests/datatype/tests_Date.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    auto lit5 = Literal("2022-03-02", type_iri);
+    CHECK(lit5.value<datatypes::xsd::Date>() == value);
+
+    // Fill in values for Date
+    tm.tm_sec = 0; // seconds of minutes from 0 to 61
+    tm.tm_min = 0; // minutes of hour from 0 to 59
+    tm.tm_hour = 0; // hours of day from 0 to 24
+    tm.tm_year = 2022 - 1900; // year since 1900
+    tm.tm_mon = 6 - 1; // month of year from 0 to 11
+    tm.tm_mday = 31;
```
Comment (nkaralis) [inline tests/datatype/tests_Date.cpp]: You need to document why this is equal to July 1st.

File: src/rdf4cpp/rdf/datatypes/xsd/DateTime.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+template<>
+struct DatatypeMapping<xsd_dateTime> {
+    using cpp_datatype = time_t;
+};
+
+/**
+ * Specialisation of from_string template function.
+ */
+template<>
+inline LiteralDatatypeImpl<xsd_dateTime>::cpp_type LiteralDatatypeImpl<xsd_dateTime>::from_string(std::string_view s) {
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/DateTime.hpp:35]: As in Date, we need to deal with the "day-of-month values" constraint and the timezone fragment.

File: tests/datatype/tests_DateTimeStamp.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    // Fill in values for DateTime
+    tm.tm_sec = 61; // seconds of minutes from 0 to 61
+    tm.tm_min = 20; // minutes of hour from 0 to 59
+    tm.tm_hour = 5; // hours of day from 0 to 24
+    tm.tm_year = 2022 - 1900; // year since 1900
+    tm.tm_mon = 4 - 1; // month of year from 0 to 11
+    tm.tm_mday = 22; // date of the month
+    tm.tm_isdst = -1; // value should be set even if not used
+    tm.tm_zone = "CEST"; //timezone
```
Comment (nkaralis) [inline tests/datatype/tests_DateTimeStamp.cpp:59]: Test with other timezones as well.

File: src/rdf4cpp/rdf/datatypes/xsd/Time.hpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * Specialisation of from_string template function.
+ */
+template<>
+inline LiteralDatatypeImpl<xsd_time>::cpp_type LiteralDatatypeImpl<xsd_time>::from_string(std::string_view s) {
+
+    const std::regex time_regex(R"((([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\.[0-9]+)?|(24:00:00(\.0+)?))(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?)");
+
+    if (std::regex_match(s.data(), time_regex)) {
+        struct tm tm {};
+        strptime(s.data(), "%H:%M:%S", &tm);
```
Comment (nkaralis) [inline src/rdf4cpp/rdf/datatypes/xsd/Time.hpp:42]: We need to handle the optional timezone fragment.

File: tests/datatype/tests_Time.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+
+#include <doctest/doctest.h>
+#include <rdf4cpp/rdf.hpp>
+
+using namespace rdf4cpp::rdf;
+
+TEST_CASE("Datatype Time") {
```
Comment (nkaralis) [inline tests/datatype/tests_Time.cpp:8]: We need to test the values '00:00:00' and '24:00:00'. According to the standard, the lexical mapping maps both representations to the same value.

**Review by nkaralis — CHANGES_REQUESTED** [review]

File: tests/datatype/tests_Time.cpp (hunk 8 above)
Comment (bigerl) [reply tests/datatype/tests_Time.cpp:8]: not done yet

**Comment by liss-h** [comment]
Superceeded by #179

**Pull request closed without merging**